motionplan/collision.go

```diff
@@ -84,140 +74,118 @@ func NewObjectCollisionEntities(geometries map[string]spatial.Geometry) (*Object
+// 		return 1, nil
+// 	}
+// 	return -1, nil
+// }
+
+// func (sce spaceCollisionEntities) reportCollisions(distances []float64) []int {
+// 	collisionIndices := make([]int, 0)
+// 	for i := range distances {
+// 		if distances[i] >= spatial.CollisionBuffer {
+// 			collisionIndices = append(collisionIndices, i)
+// 		} else {
+// 			return []int{}
+// 		}
+// 	}
+// 	return collisionIndices
+// }
+
+// TODO: comments
+type entityGraph struct {
+	x, y *collisionEntities
+
+	distances [][]float64
+
+	// triangular is a bool that describes if the adjacencies matrix is triangular, which will be the case when x ==y
+	triangular bool
 
-// NewSpaceCollisionEntities is a constructor for spaceCollisionEntities.
-func NewSpaceCollisionEntities(geometries map[string]spatial.Geometry) (CollisionEntities, error) {
-	entities, err := NewObjectCollisionEntities(geometries)
-	return spaceCollisionEntities{entities}, err
+	reportDistances bool
 }
 
-func (sce spaceCollisionEntities) checkCollision(key, test *collisionEntity, reportDistances bool) (float64, error) {
-	encompassed, err := key.geometry.EncompassedBy(test.geometry)
-	if err != nil {
-		return math.NaN(), err
-	}
-	// TODO(rb): EncompassedBy should also report distance required to resolve the collision
-	if !encompassed {
-		return 1, nil
+func newEntityGraph(x, y *collisionEntities, reportDistances bool) *entityGraph {
+	return &entityGraph{
+		x:               x,
+		y:               y,
+		distances:       make([][]float64, x.count()),
+		reportDistances: reportDistances,
+		triangular:      x == y,
 	}
-	return -1, nil
 }
 
-func (sce spaceCollisionEntities) reportCollisions(distances []float64) []int {
-	collisionIndices := make([]int, 0)
-	for i := range distances {
-		if distances[i] >= spatial.CollisionBuffer {
-			collisionIndices = append(collisionIndices, i)
-		} else {
-			return []int{}
-		}
-	}
-	return collisionIndices
+func (cg *entityGraph) getIndices(xName, yName string) (int, int, bool) {
+	i := cg.x.indexFromName(xName)
+	j := cg.y.indexFromName(yName)
+	return i, j, i >= 0 && j >= 0
 }
 
-// collisionGraph is an implementation of an undirected graph used to track collisions between two set of CollisionEntities.
 type collisionGraph struct {
-	// key CollisionEntities
-	key *ObjectCollisionEntities
-
-	// test CollisionEntities are the set of CollisionEntities from which the collisionGraph takes its
-	// collisionCheckFn and collisionReportFn functions to check and report collisions between the
-	// test CollisionEntities and key Collision Entities
-	test CollisionEntities
-
-	// adjacencies is 2D array encoding edges between collisionEntiies in the collisionGraph.
-	// if adjacencies[i][j] >= 0 this corresponds to an edge between the entities at key[i] and test[j]
-	adjacencies [][]float64
-
-	// triangular is a bool that describes if the adjacencies matrix is triangular, which will be the case when key == test
-	triangular bool
-
-	reportDistances bool
+	entityGraph
```

Comment:
[q] This is structured so that future graphs representative of different kinds of interactions can use the common `entityGraph` format?

Reply:
Yes the idea is that the core functionality of these graphs will live behind `entityGraph` and the different kinds of interactions will each inherit from them